File: extension/src/shared/types/page.ts
export interface ExtractedTextItem {
  item_id: string;
  text: string;
  tag_name: string;
  page_url?: string;
  meta?: Record<string, unknown>;
}

export interface ExtractedImageItem {
  item_id: string;
  src: string;
  alt_text?: string;
  width: number;
  height: number;
  tag_name: string;
  page_url?: string;
  meta?: Record<string, unknown>;
}

export interface ExtractedPromoItem {
  item_id: string;
  text: string;
  tag_name: string;
  role?: string;
  class_name?: string;
  page_url?: string;
  meta?: Record<string, unknown>;
}

export interface ExtractedHeadingItem {
  item_id: string;
  text: string;
  level: number;
  page_url?: string;
  meta?: Record<string, unknown>;
}

export interface ExtractedPage {
  url: string;
  title: string;
  text_items: ExtractedTextItem[];
  image_items: ExtractedImageItem[];
  promo_items: ExtractedPromoItem[];
  headings: ExtractedHeadingItem[];
}

File: backend/app/presentation/schemas/page.py
from typing import Any

from pydantic import BaseModel, Field

from app.domain.entities.page_analysis import PageAnalysis
from app.domain.entities.user_preferences import UserPreferences


class PromoItemInput(BaseModel):
    item_id: str
    text: str
    tag_name: str | None = None
    role: str | None = None
    class_name: str | None = None
    page_url: str | None = None
    meta: dict[str, Any] = Field(default_factory=dict)


class HeadingItemInput(BaseModel):
    item_id: str
    text: str
    level: int = 1
    page_url: str | None = None
    meta: dict[str, Any] = Field(default_factory=dict)


class PageAnalysisRequest(BaseModel):
    url: str
    title: str
    text_items: list[dict] = Field(default_factory=list)
    image_items: list[dict] = Field(default_factory=list)
    promo_items: list[dict] = Field(default_factory=list)
    headings: list[dict] = Field(default_factory=list)
    preferences: UserPreferences | None = None


class PageAnalysisResponse(BaseModel):
    analysis: PageAnalysis
